### PR TITLE
Combine Web and CLI Authentication Flows

### PR DESCRIPTION
**Pull request by jhunt**
We can now support both CLI and Web Auth Provider flows.
Each authentication provider now has two entry point URIs,
once for web users and another for CLI users (by way of a
web browser, to make it all work).  As people enter one way
or another, they get a `via` cookie that tracks their particular
entry point, so we can handle them on redirect properly.

On redirect, web users continue to get a `Set-Cookie` header
in their final redirect back to SHIELD's `/` page.  CLI users
will instead get a new page, `#!/cliauth` that presents them
with their newly-minted session ID, and instructs them to
copy / paste it into their SHIELD CLI (none of the receiving
end work has been done yet, an is out of scope for this PR).

### CODE REVIEW

File: core/github_auth.go [hunk 1]
```diff
@@ -29,130 +30,111 @@ type GithubAuthProvider struct {
 		} `json:"rights"`
 	} `json:"mapping"`
 
-	Name       string
-	Identifier string
-	Usage      string
-	core       *Core
+	Name  string
+	Usage string
```
Comment (thomasmmitchell) [inline core/github_auth.go]: What are we still using Usage for?

File: core/github_auth.go [hunk 2]
```diff
@@ -29,130 +30,111 @@ type GithubAuthProvider struct {
...
+	core  *Core
 }
 
-func (gh *GithubAuthProvider) DisplayName() string {
-	return gh.Name
-}
+func (p *GithubAuthProvider) Configure(raw map[interface{}]interface{}) error {
+	p.Type = "github"
```
Comment (thomasmmitchell) [inline core/github_auth.go]: Is this redundant? This is set in auth.go with the rest of the AuthProviderBase

**Comment by thomasmmitchell** [comment]
How are we determining that Token shouldn't be shown in the web ui?

File: core/api.go [hunk 3]
```diff
@@ -28,8 +28,55 @@ func (core *Core) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+			w.WriteHeader(404)
+			fmt.Fprintf(w, "couldn't find that auth provider...: %s", err)
+			return
+		}
+
+		if m[2] == "redir" {
+			via := "web"
+			if cookie, err := req.Cookie("via"); err == nil {
+				via = cookie.Value
+			}
```
Comment (thomasmmitchell) [inline core/api.go]: Should we kill the via Cookie at this point

File: core/v2.go [hunk 4]
```diff
@@ -225,28 +221,34 @@ func (core *Core) v2API() *route.Router {
...
+		typ := r.Param("for", "cli")
 		for _, auth := range core.auth {
-			if usage == "" || auth.Usage == usage {
-				l = append(l, v2AuthProvider{
-					Name:       auth.Name,
-					Identifier: auth.Identifier,
-					Type:       auth.Backend,
-					Usage:      auth.Usage,
-				})
+			if auth.Backend == "token" && typ != "cli" {
```
Comment (jhunt) [inline core/v2.go]: @thomasmmitchell this is where we hide the token backend from the list of available auth providers, unless you want CLI providers.

File: core/api.go [hunk 5]
```diff
@@ -28,8 +28,55 @@ func (core *Core) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+				fmt.Fprintf(w, "The authentication process broke down\n")
+				w.WriteHeader(500)
+			}
+
+			session, err := core.createSession(user)
+			if err != nil {
+				log.Errorf("failed to create a session for user %s@%s: %s", user.Account, user.Backend, err)
+				w.Header().Set("Location", "/")
+			} else if via == "cli" {
+				w.Header().Set("Location", fmt.Sprintf("/#!/cliauth:s:%s", session.UUID.String()))
```
Comment (thomasmmitchell) [inline core/api.go]: Can you redirect to an address with a query param?

File: core/github_auth.go (hunk 1 above)
Comment (jhunt) [reply core/github_auth.go]: Oops.  Missed that one.

File: core/api.go (hunk 3 above)
Comment (jhunt) [reply core/api.go]: Thought about it, but it doesn't hurt anything, and it makes it easier to troubleshoot ex post facto.

File: core/api.go (hunk 5 above)
Comment (jhunt) [reply core/api.go]: Yes, but we're redirecting to a URI with a fragment identifier.

**Comment by jhunt** [comment]
@thomasmmitchell feedback addressed, can you take another look?

**Pull request merged**